pxtcompiler/emitter/languageservice.ts

```diff
@@ -0,0 +1,650 @@
+        }
+    }
+
+    export function getTsSymbolFromPxtSymbol(pxtSym: SymbolInfo, location: ts.Node, meaning: SymbolFlags): ts.Symbol | null {
+        const checker = service && service.getProgram().getTypeChecker();
+        if (!checker)
+            return null
+        const tsSymbols = checker.getSymbolsInScope(location, meaning)
+        for (let tsSym of tsSymbols) {
+            if (tsSym.escapedName.toString() === pxtSym.qName)
+                return tsSym
+        }
+        return null
+    }
+
+
+
+    export function getDefaultEnumValue(t: Type, python: boolean): string {
+        // Note: AFAIK this is NOT guranteed to get the same default as you get in
+        // blocks. That being said, it should get the first declared value. Only way
+        // to guarantee an API has the same default in blocks and in TS is to actually
+        // set a default on the parameter in its comment attributes
+        const checker = service && service.getProgram().getTypeChecker();
+        const members = getEnumMembers(checker, t)
+        for (const member of members) {
+            if (member.name.kind === SK.Identifier) {
+                const fullName = enumMemberToQName(checker, member)
+                const pxtSym = lastApiInfo.apis.byQName[fullName]
+                if (pxtSym) {
+                    if (pxtSym.attributes.alias)
+                        // use pyAlias if python; or default to alias
+                        return (python && pxtSym.attributes.pyAlias) || pxtSym.attributes.alias; // prefer alias
+                    return python ? pxtSym.pyQName : pxtSym.qName
+                }
+                else
+                    return fullName
+            }
+        }
+
+        return "0";
+    }
+
+    export function getCompletions(v: OpArg) {
+        const { fileName, fileContent, position, wordStartPos, wordEndPos, runtime } = v
+        let src: string = fileContent
+        if (fileContent) {
+            host.setFile(fileName, fileContent);
+        }
+
+        const tsFilename = filenameWithExtension(fileName, "ts");
+
+        const span: PosSpan = { startPos: wordStartPos, endPos: wordEndPos }
+
+        const isPython = /\.py$/.test(fileName);
+
+        const r: CompletionInfo = {
+            entries: [],
+            isMemberCompletion: false,
+            isNewIdentifierLocation: true,
+            isTypeLocation: false,
+            namespace: [],
+        };
+
+        // get line text
+        let lastNl = src.lastIndexOf("\n", position - 1)
+        lastNl = Math.max(0, lastNl)
+        const lineText = src.substring(lastNl + 1, position)
+
+        // are we on a line comment, if so don't show completions
+        // NOTE: multi-line comments and string literals are handled
+        //  later as they require parsing
+        const lineCommentStr = isPython ? "#" : "//"
+        if (lineText.trim().startsWith(lineCommentStr)) {
+            return r;
+        }
```

Comment:
This is new. This aborts if we're on a comment line

Reply:
Checking my understanding here; this won't bail out on the following,

```typescript
console.log("got here") // I got here!!
```

but that would get handled in the commentMap section below, it just doesn't bail out early.

Reply:
Correct, the above is just a quick abort for perf reasons. It could actually be made more general to capture end-of-line comments as well.